### PR TITLE
feat(metadata service): Introducing Platform Events 

### PR DESCRIPTION
**Pull request by jjoyce0510**
**Summary** 
This PR introduces a new concept into the Metadata Service (GMS) of **Platform Event**. A Platform Event is a custom type of business (e.g. semantic) event that can be registered via the entity-registry. Platform Events are effectively platform-internal kafka events that can be emitted using the "producePlatformEvent" API of an EntityClient. When producing an event, its "name" and "payload" must be provided. 

In contrast to MCP + MCL events, Platform Events do not imply any storage other than the event queue (e.g. Kafka). Platform Events are not automatically written to elastic or MySQL. Instead, it is assumed that they will be consumed from internal or external consumers of the Platform Event message stream. 

Following a similar precedent to how aspects are serialized in generic fashion, Platform Events are serialized using the concept of GenericPayload inside of an envelope record called PlatformEvent, meaning that we can add and produce arbitrary platform events on a single topic (PlatformEvent_v1) without a schema registry change. 

**Changes**
- New type of event topic in GMS: Platform Event 
- Extend Entity Registry related components to support configuring, parsing, validating, and serving details about "EventSpecs" 
- APIs for producing platform events
- Adding a "platform event consumer" process inside of MAE-consumer job. For now it's a no-op. 
- Updating entity registry related tests to verify platform events
- Generalize GenericAspectUtils to GenericRecordUtils, since some logic is common between entity aspects and event payloads. 
- Re-brand EntityEventProducer -> EventProducer to generalize. 
- Re-brand EntityKafkaEventProducer -> KafkaEventProducer to generalize. 

In coming PRs, we will add a few sample PlatformEvents which will more fully demonstrate their power! 

## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: docker/kafka-setup/kafka-setup.sh [hunk 1]
```diff
@@ -24,12 +24,14 @@ cub kafka-ready -c $CONNECTION_PROPERTIES_PATH -b $KAFKA_BOOTSTRAP_SERVER 1 60
...
+
 # Set retention to 90 days
 kafka-topics.sh --create --if-not-exists --command-config $CONNECTION_PROPERTIES_PATH --bootstrap-server $KAFKA_BOOTSTRAP_SERVER --partitions $PARTITIONS --replication-factor $REPLICATION_FACTOR --config retention.ms=7776000000 --topic $METADATA_CHANGE_LOG_TIMESERIES_TOPIC
 kafka-topics.sh --create --if-not-exists --command-config $CONNECTION_PROPERTIES_PATH --bootstrap-server $KAFKA_BOOTSTRAP_SERVER --partitions $PARTITIONS --replication-factor $REPLICATION_FACTOR --topic $METADATA_CHANGE_PROPOSAL_TOPIC
 kafka-topics.sh --create --if-not-exists --command-config $CONNECTION_PROPERTIES_PATH --bootstrap-server $KAFKA_BOOTSTRAP_SERVER --partitions $PARTITIONS --replication-factor $REPLICATION_FACTOR --topic $FAILED_METADATA_CHANGE_PROPOSAL_TOPIC
+kafka-topics.sh --create --if-not-exists --command-config $CONNECTION_PROPERTIES_PATH --bootstrap-server $KAFKA_BOOTSTRAP_SERVER --partitions $PARTITIONS --replication-factor $REPLICATION_FACTOR --topic $PLATFORM_EVENT_TOPIC_NAME
```
Comment (dexter-mh-lee) [inline docker/kafka-setup/kafka-setup.sh:34]: why do we need to create this two times?

Comment (jjoyce0510) [reply docker/kafka-setup/kafka-setup.sh:34]: wym?

File: entity-registry/src/main/java/com/linkedin/metadata/models/DefaultEventSpec.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
+package com.linkedin.metadata.models;
+
+import com.linkedin.data.schema.RecordDataSchema;
+import com.linkedin.metadata.models.annotation.EventAnnotation;
+import javax.annotation.Nonnull;
+
+
+public class DefaultEventSpec implements EventSpec {
```
Comment (dexter-mh-lee) [inline entity-registry/src/main/java/com/linkedin/metadata/models/DefaultEventSpec.java:8]: We can do something like below to remove boilerplate code. (Note you need to remove _ from variable name to make it work)
```
@RequiredArgsConstructor
@Getter
public class DefaultEventSpec implements EventSpec {
  private final String name;
  private final EventAnnotation eventAnnotation;
  private final RecordDataSchema pegasusSchema;
}
```

File: docker/kafka-setup/kafka-setup.sh (hunk 1 above)
Comment (dexter-mh-lee) [reply docker/kafka-setup/kafka-setup.sh:34]: I see the same line above on line 28? are these doing different things?

File: entity-registry/src/main/java/com/linkedin/metadata/models/registry/config/Event.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.NoArgsConstructor;
+import lombok.Value;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+@Value
+@NoArgsConstructor(force = true, access = AccessLevel.PRIVATE)
+@AllArgsConstructor
```
Comment (dexter-mh-lee) [inline entity-registry/src/main/java/com/linkedin/metadata/models/registry/config/Event.java:11]: I believe @Value already sets this annotation

File: entity-registry/src/main/java/com/linkedin/metadata/models/registry/config/Event.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package com.linkedin.metadata.models.registry.config;
+
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.NoArgsConstructor;
+import lombok.Value;
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+@Value
+@NoArgsConstructor(force = true, access = AccessLevel.PRIVATE)
```
Comment (dexter-mh-lee) [inline entity-registry/src/main/java/com/linkedin/metadata/models/registry/config/Event.java:10]: do we need this? thought by having a constructor it removes the default constructor

File: docker/datahub-gms/env/docker.env [hunk 5]
```diff
@@ -17,6 +17,7 @@ GRAPH_SERVICE_IMPL=neo4j
 ENTITY_REGISTRY_CONFIG_PATH=/datahub/datahub-gms/resources/entity-registry.yml
 MAE_CONSUMER_ENABLED=true
 MCE_CONSUMER_ENABLED=true
+PE_CONSUMER_ENABLED=true
```
Comment (dexter-mh-lee) [inline docker/datahub-gms/env/docker.env:20]: do we need to add these changes to mae-consumer's docker.env as well?

Comment (dexter-mh-lee) [reply docker/datahub-gms/env/docker.env:20]: you also need to run the generate quickstart script

Comment (jjoyce0510) [reply docker/datahub-gms/env/docker.env:20]: Oh good call - yes!

File: docker/kafka-setup/kafka-setup.sh (hunk 1 above)
Comment (jjoyce0510) [reply docker/kafka-setup/kafka-setup.sh:34]: oh yeah weird

**Review by dexter-mh-lee — APPROVED** [review]

**Pull request merged**